### ISSUE TITLE
🐛 BUG: Vue components cannot be nested

### ISSUE THREAD

**Issue by gqio** [snowpackjs/astro#182]
Hi team, 

What I would like to do, is being able to nest some components and reuse only big component like a card composed by several components underneath

I tried to replicate my issue with a minimal code just here (cc @matthewp )

This one does not work
https://github.com/gqio/astro-local-project-example/commit/027b36772a9bbcfb4707fecda3d5cfc1fd3674aa

But this one works
https://github.com/gqio/astro-local-project-example/commit/3c9a6db8a69d80cd29f6fd137d2f89a988502d21

Thanks Team

**Comment by natemoo-re** [comment]
Just bumping this thread to let everyone know I'm looking at this now.

**Comment by natemoo-re** [comment]
Okay, this has been a total rabbit hole. Here are some potentially related issues and commits.

https://github.com/vuejs/vue-next/issues/2863

https://github.com/vuejs/vue-next/commit/8c3c14a0ff02bffbc37e1b069d4ff3c7e086a1d5#diff-561be13280b90fe64842cc767742d3ade26e09edc35b7db1ea221c3a68297e63R830

**Comment by TimWeidner** [comment]
> Okay, this has been a total rabbit hole. Here are some potentially related issues and commits. I'm suspecting that this is an upstream issue, but I'm not sure how to confirm that suspicion.
> 
> [vuejs/vue-next#2863](https://github.com/vuejs/vue-next/issues/2863)
> 
> [vuejs/vue-next@8c3c14a#diff-561be13280b90fe64842cc767742d3ade26e09edc35b7db1ea221c3a68297e63R830](https://github.com/vuejs/vue-next/commit/8c3c14a0ff02bffbc37e1b069d4ff3c7e086a1d5#diff-561be13280b90fe64842cc767742d3ade26e09edc35b7db1ea221c3a68297e63R830)

The linked repository workes perfectly fine with vue 3.1.2 though, What is your issue with the linked commit, @natemoo-re ?

**Comment by natemoo-re** [comment]
The linked example does work correctly in `vue@3.1.2`. But even when I upgrade our Vue renderer to `3.1.2`, I hit (seemingly) the same issue. The difference in our case is that the components are not all at the top-level of the `app`, they are recursive.

```js
const B = { template: `<div id="b">B</div>` };
const A = { components: { B }, template: `<div id="a"><B /></div>` };

createApp({
  components: { A },
  render() {
    return h(A, {});
  }
})
```

**Comment by natemoo-re** [comment]
For more context, I see the following in the console, which is part of the problem that was reported.

```
[Vue warn]: resolveComponent can only be used in render() or setup().
```

This [workaround](https://github.com/lovetingyuan/vue3-ssr-dynamic-component-bug/blob/683579d992de31f335971c76e17c92c89fe6004a/src/App.vue#L24) was suggested but doesn't work in this case.

**Comment by TimWeidner** [comment]
I do actually have some kind of idea what could cause this. I might be completely wrong, but I'll try this - either late today or early tomorrow.

**Comment by TimWeidner** [comment]
Hey, so basically I am not sure what exactly causes this, but I think that vue might not see the nested component as a actual component (is the astro-fragment element injected between the two components as well?)

**Comment by Janne252** [comment]
I ran into this issue as well. It seems that any custom element, even non-registered inside of a Vue component's template causes 
 the `[Vue warn]: resolveComponent can only be used in render() or setup().` warning and the fatal error `[executing astro] TypeError: Invalid value used as weak map key`.
 
 ```vue
 <template>
    <what-ever />
</template>

<script>
export default {}
</script>

<style>
</style>
```

**Comment by Janne252** [comment]
It appears that I can get around this issue by using dynamic components: `<component :is="SomeComponent" />`

```vue
<template>
    <ul>
        <li v-for="post in items" :key="post.url">
            <a :href="post.url">{{ post.title }}</a>
            <component :is="Tags" :items="post.tags" />
        </li>
    </ul>
</template>

<script>
import Tags from './Tags.vue';
export default {
    data() {
        return {
            Tags,      
        }
    },
    props: {
        items: {
            type: Array,
            default: () => new Array()
        }
    }
}
</script>

<style>

</style>
```

**Comment by matthewp** [comment]
Fixed in https://github.com/snowpackjs/astro/pull/858

**Comment by matthewp** [comment]
Reopening as we had to revert the change because it was causing issues with client-side Vue.

**Comment by Igloczek** [comment]
I get to the point it actually works, this time on both server and client side.

On Astro side I removed this https://github.com/snowpackjs/astro/blob/main/packages/astro/src/external.ts#L21 to let `@vue/server-renderer` be not taken as external source, and be part of the Snowpack cache.

On Vue side, I made a ESM version of `@vue/server-renderer` , but I had to to change one potentialy breaking thing there, so hope it will be accepted or there is other way of handling that :fingers_crossed:

PRs in progress...

**Comment by Igloczek** [comment]
Vue stuff merged, we just need to wait for release. 

@matthewp do we need to wait for a stable release, or beta will be fine?

**Comment by matthewp** [comment]
@Igloczek if it's a non-breaking release (not major) then I think a beta is ok.

**Comment by matthewp** [comment]
Ok think it's fixed for real this time! In #924
